README.md
# Project 2

    u: Alexander Scheel
    e: <scheel@iastate.edu>
    l: BSD: 2-clause

## Overview
This is my final project for COM S 327. The goal of this project are as follows:

1. Implement the Blowfish cryptographic primitive
2. Implement block modes on top of the primitive
3. Demonstrate attacks against the various block modes

For clarity of data typing, this project uses C++-11 for the availability of
uint32_t and uint8_t; many other implementations use "unsigned long", which
is not compatible across operating systems. In particular, a number of systems
define this as an 64-bit type, but it is necessary to have it as a 32-bit type
for the implementation of blowfish.

The source for the primitive and various abstracted block modes are in the
/src directory. The two currently implemented block modes are ECB and CBC.
Eventually, CTR mode (stream cipher) might be implemented. For the block
modes, the block_(en/de)crypt functions have been overloaded to support both
the encryption of char* style strings and std::string style strings.

The source for the testing of this implementation are located in the /test
directory. The test vectors are those given by Schneier, attributed to 
Young. This is the defacto set of test vectors, matched by Go's and LibreSSL's
Blowfish implementations, along with many other reference implementations.

Finally, the source for the attacks against the various block modes are 
located in the /main directory. As of the time of submission, there are
two attacks implemented: appended-string decryption with unknown key, 
and a mock privilege escalation attack, as if given an encrypted cookie. These
attacks are modeled after Cryptopals challenges 12 and 13. 


Documentation of Blowfish, block mode APIs, and of the attacks are located
in the /docs directory.


## Running
To compile and run the tests to ensure validity of the implementation:

    make all
    ./bin/tests

The attacks are implemented in main/main.cc; to run:

    make all
    ./bin/blowfish


CHANGELOG
commit + f449c24	refs/heads/master (HEAD -> refs/heads/master, refs/remotes/origin/master, refs/remotes/origin/HEAD)
log size 115
Author: Alexander Scheel <alexander.m.scheel@gmail.com>
Date:   6 minutes ago

    Added description of function_f

 docs/src/blowfish.txt | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

commit + 7560d9c f449c24	refs/heads/master
log size 132
Author: Alexander Scheel <alexander.m.scheel@gmail.com>
Date:   3 hours ago

    Unrolled loop; documentation on blowfish algorithm

 CHANGELOG             |  71 +++++++++++++++++++---------
 Makefile              |   1 +
 docs/src/blowfish.txt |  47 ++++++++++++++++++-
 src/blowfish.cc       | 127 ++++++++++++++++++++++++++++++++++----------------
 4 files changed, 180 insertions(+), 66 deletions(-)

commit + a3a70b6 7560d9c	refs/heads/master
log size 104
Author: Alexander Scheel <alexander.m.scheel@gmail.com>
Date:   21 hours ago

    Started documentation

 docs/src/blowfish.txt  | 18 +++++++++++
 docs/tests/vectors.txt | 85 ++++++++++++++++++++++++++++++++++++++++++++++++++
 docs/vectors.txt       | 85 --------------------------------------------------
 3 files changed, 103 insertions(+), 85 deletions(-)

commit + 12bdbe6 a3a70b6	refs/heads/master
log size 100
Author: Alexander Scheel <alexander.m.scheel@gmail.com>
Date:   22 hours ago

    Working CBC mode.

 CHANGELOG           |  26 +++++++++----
 main/main.cc        |  11 ++++--
 src/blowfish_cbc.cc | 109 ++++++++++++++++++++++++++++++++++++++++++++++++++++
 src/blowfish_cbc.hh |  23 +++++++++++
 4 files changed, 158 insertions(+), 11 deletions(-)

commit + 6812290 12bdbe6	refs/heads/master
log size 106
Author: Alexander Scheel <alexander.m.scheel@gmail.com>
Date:   22 hours ago

    Finished ECB block mode

 CHANGELOG           | 43 +++++++++++++++++++++--------------
 main/main.cc        | 29 ++++++++++++++++--------
 src/blowfish_ecb.cc | 64 +++++++++++++++++++++++++++++++++--------------------
 3 files changed, 87 insertions(+), 49 deletions(-)

commit + 7a94892 6812290	refs/heads/master
log size 112
Author: Alexander Scheel <alexander.m.scheel@gmail.com>
Date:   23 hours ago

    Fixed gitignore for bin/tests

 .gitignore |  1 +
 CHANGELOG  | 42 +++++++++++++++++++++++++++++++-----------
 2 files changed, 32 insertions(+), 11 deletions(-)

commit + ff4f5ec 7a94892	refs/heads/master
log size 129
Author: Alexander Scheel <alexander.m.scheel@gmail.com>
Date:   23 hours ago

    Renamed headers to .hh; started ECB block mode

 CHANGELOG           |  32 ++++++++++++++-----
 Makefile            |   2 +-
 bin/tests           | Bin 59976 -> 0 bytes
 main/main.cc        |  23 +++++++++-----
 src/blowfish.cc     |  87 +++++++++++++++++++---------------------------------
 src/blowfish.h      |  35 ---------------------
 src/blowfish.hh     |  35 +++++++++++++++++++++
 src/blowfish_ecb.cc |  65 +++++++++++++++++++++++++++++++++++++++
 src/blowfish_ecb.hh |  23 ++++++++++++++
 tests/schneier.cc   |   2 +-
 tests/vectors.cc    |   2 --
 11 files changed, 196 insertions(+), 110 deletions(-)

commit + d26619f ff4f5ec	refs/heads/master
log size 114
Author: Alexander Scheel <alexander.m.scheel@gmail.com>
Date:   24 hours ago

    Additional test cases: set_keys

 CHANGELOG         |  25 ++++++++++++----
 bin/tests         | Bin 0 -> 59976 bytes
 docs/tests.txt    |  34 ----------------------
 docs/vectors.txt  |  85 ++++++++++++++++++++++++++++++++++++++++++++++++++++++
 src/blowfish.cc   |  16 ++++------
 src/blowfish.h    |  12 ++++----
 tests/schneier.cc |  65 +++++++++++++++++++++++++++++++++++++++++
 7 files changed, 181 insertions(+), 56 deletions(-)

commit + 1c0a4e8 d26619f	refs/heads/master
log size 108
Author: Alexander Scheel <alexander.m.scheel@gmail.com>
Date:   24 hours ago

    Automated build of tests.

 CHANGELOG         |  28 +++++--
 Makefile          |  20 +++--
 main/main.cc      |  39 ++++++++++
 src/main.cc       |  39 ----------
 tests/schneier.cc | 213 ++++++++++++++++++++++++++++--------------------------
 tests/vectors.cc  |  16 +++-
 6 files changed, 197 insertions(+), 158 deletions(-)

commit + a8ef3f8 1c0a4e8	refs/heads/master
log size 117
Author: Alexander Scheel <alexander.m.scheel@gmail.com>
Date:   24 hours ago

    Passing test cases: Bruce Schneier

 CHANGELOG         |  31 ++++++++----
 docs/tests.txt    |  34 +++++++++++++
 src/blowfish.cc   |  18 +++----
 src/blowfish.h    |   4 +-
 src/main.cc       |  10 +---
 tests/schneier.cc | 140 ++++++++++++++++++++++++++++++++++++++++++++++++++++++
 tests/vectors.cc  |  10 ++++
 7 files changed, 219 insertions(+), 28 deletions(-)

commit + 9858898 a8ef3f8	refs/heads/master
log size 102
Author: Alexander Scheel <alexander.m.scheel@gmail.com>
Date:   26 hours ago

    Fixed bin directory

 .gitignore |  2 +-
 CHANGELOG  | 20 ++++++++++++++++----
 Makefile   |  2 +-
 bin/.keep  |  0
 4 files changed, 18 insertions(+), 6 deletions(-)

commit + 5385496 9858898	refs/heads/master
log size 107
Author: Alexander Scheel <alexander.m.scheel@gmail.com>
Date:   27 hours ago

    Fixed key initialization

 CHANGELOG       |  27 ++++++++++----
 src/blowfish.cc | 108 +++++++++++++++++++++++++-------------------------------
 src/main.cc     |  20 ++++-------
 3 files changed, 74 insertions(+), 81 deletions(-)

commit + de0ba9f 5385496	refs/heads/master
log size 100
Author: Alexander Scheel <alexander.m.scheel@gmail.com>
Date:   29 hours ago

    Fixed decryption.

 CHANGELOG       |  16 ++++-
 src/blowfish.cc | 181 +++++++++++++++++++++++++++++++-------------------------
 src/blowfish.h  |  12 ++--
 src/main.cc     |  66 ++++++++++++---------
 4 files changed, 155 insertions(+), 120 deletions(-)

commit + e05dc03 de0ba9f	refs/heads/master
log size 116
Author: Alexander Scheel <alexander.m.scheel@gmail.com>
Date:   2 days ago

    Gitignore, blowfish encrypt/decrypt

 .gitignore      |  43 ++++++++++++++
 CHANGELOG       |  21 ++++++-

tests/vectors.cc
#include <iostream>

extern bool schneier_ecb_test();
extern bool schneier_set_key_test();

int main(void)
{
    if (!schneier_ecb_test()) {
        std::cout << "Tests failed" << std::endl;
        return 1;
    }

    if (!schneier_set_key_test()) {
        std::cout << "Tests failed" << std::endl;
        return 1;
    }
    return 0;
}


tests/schneier.cc
#include "../src/blowfish.hh"
#include <cstdlib>
#include <iostream>

bool schneier_ecb_test()
{
    char keys[34][8] = {{(char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00},
        {(char) 0xFF, (char) 0xFF, (char) 0xFF, (char) 0xFF, (char) 0xFF, (char) 0xFF, (char) 0xFF, (char) 0xFF},
        {(char) 0x30, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00},
        {(char) 0x11, (char) 0x11, (char) 0x11, (char) 0x11, (char) 0x11, (char) 0x11, (char) 0x11, (char) 0x11},
        {(char) 0x01, (char) 0x23, (char) 0x45, (char) 0x67, (char) 0x89, (char) 0xAB, (char) 0xCD, (char) 0xEF},
        {(char) 0x11, (char) 0x11, (char) 0x11, (char) 0x11, (char) 0x11, (char) 0x11, (char) 0x11, (char) 0x11},
        {(char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00},
        {(char) 0xFE, (char) 0xDC, (char) 0xBA, (char) 0x98, (char) 0x76, (char) 0x54, (char) 0x32, (char) 0x10},
        {(char) 0x7C, (char) 0xA1, (char) 0x10, (char) 0x45, (char) 0x4A, (char) 0x1A, (char) 0x6E, (char) 0x57},
        {(char) 0x01, (char) 0x31, (char) 0xD9, (char) 0x61, (char) 0x9D, (char) 0xC1, (char) 0x37, (char) 0x6E},
        {(char) 0x07, (char) 0xA1, (char) 0x13, (char) 0x3E, (char) 0x4A, (char) 0x0B, (char) 0x26, (char) 0x86},
        {(char) 0x38, (char) 0x49, (char) 0x67, (char) 0x4C, (char) 0x26, (char) 0x02, (char) 0x31, (char) 0x9E},
        {(char) 0x04, (char) 0xB9, (char) 0x15, (char) 0xBA, (char) 0x43, (char) 0xFE, (char) 0xB5, (char) 0xB6},
        {(char) 0x01, (char) 0x13, (char) 0xB9, (char) 0x70, (char) 0xFD, (char) 0x34, (char) 0xF2, (char) 0xCE},
        {(char) 0x01, (char) 0x70, (char) 0xF1, (char) 0x75, (char) 0x46, (char) 0x8F, (char) 0xB5, (char) 0xE6},
        {(char) 0x43, (char) 0x29, (char) 0x7F, (char) 0xAD, (char) 0x38, (char) 0xE3, (char) 0x73, (char) 0xFE},
        {(char) 0x07, (char) 0xA7, (char) 0x13, (char) 0x70, (char) 0x45, (char) 0xDA, (char) 0x2A, (char) 0x16},
        {(char) 0x04, (char) 0x68, (char) 0x91, (char) 0x04, (char) 0xC2, (char) 0xFD, (char) 0x3B, (char) 0x2F},
        {(char) 0x37, (char) 0xD0, (char) 0x6B, (char) 0xB5, (char) 0x16, (char) 0xCB, (char) 0x75, (char) 0x46},
        {(char) 0x1F, (char) 0x08, (char) 0x26, (char) 0x0D, (char) 0x1A, (char) 0xC2, (char) 0x46, (char) 0x5E},
        {(char) 0x58, (char) 0x40, (char) 0x23, (char) 0x64, (char) 0x1A, (char) 0xBA, (char) 0x61, (char) 0x76},
        {(char) 0x02, (char) 0x58, (char) 0x16, (char) 0x16, (char) 0x46, (char) 0x29, (char) 0xB0, (char) 0x07},
        {(char) 0x49, (char) 0x79, (char) 0x3E, (char) 0xBC, (char) 0x79, (char) 0xB3, (char) 0x25, (char) 0x8F},
        {(char) 0x4F, (char) 0xB0, (char) 0x5E, (char) 0x15, (char) 0x15, (char) 0xAB, (char) 0x73, (char) 0xA7},
        {(char) 0x49, (char) 0xE9, (char) 0x5D, (char) 0x6D, (char) 0x4C, (char) 0xA2, (char) 0x29, (char) 0xBF},
        {(char) 0x01, (char) 0x83, (char) 0x10, (char) 0xDC, (char) 0x40, (char) 0x9B, (char) 0x26, (char) 0xD6},
        {(char) 0x1C, (char) 0x58, (char) 0x7F, (char) 0x1C, (char) 0x13, (char) 0x92, (char) 0x4F, (char) 0xEF},
        {(char) 0x01, (char) 0x01, (char) 0x01, (char) 0x01, (char) 0x01, (char) 0x01, (char) 0x01, (char) 0x01},
        {(char) 0x1F, (char) 0x1F, (char) 0x1F, (char) 0x1F, (char) 0x0E, (char) 0x0E, (char) 0x0E, (char) 0x0E},
        {(char) 0xE0, (char) 0xFE, (char) 0xE0, (char) 0xFE, (char) 0xF1, (char) 0xFE, (char) 0xF1, (char) 0xFE},
        {(char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00, (char) 0x00},
        {(char) 0xFF, (char) 0xFF, (char) 0xFF, (char) 0xFF, (char) 0xFF, (char) 0xFF, (char) 0xFF, (char) 0xFF},
        {(char) 0x01, (char) 0x23, (char) 0x45, (char) 0x67, (char) 0x89, (char) 0xAB, (char) 0xCD, (char) 0xEF},
        {(char) 0xFE, (char) 0xDC, (char) 0xBA, (char) 0x98, (char) 0x76, (char) 0x54, (char) 0x32, (char) 0x10}
    };

    uint32_t plaintexts[34][2] = {{0x00000000, 0x00000000},
        {0xFFFFFFFF, 0xFFFFFFFF},
        {0x10000000, 0x00000001},
        {0x11111111, 0x11111111},
        {0x11111111, 0x11111111},
        {0x01234567, 0x89ABCDEF},
        {0x00000000, 0x00000000},
        {0x01234567, 0x89ABCDEF},
        {0x01A1D6D0, 0x39776742},
        {0x5CD54CA8, 0x3DEF57DA},
        {0x0248D438, 0x06F67172},
        {0x51454B58, 0x2DDF440A},
        {0x42FD4430, 0x59577FA2},
        {0x059B5E08, 0x51CF143A},
        {0x0756D8E0, 0x774761D2},
        {0x762514B8, 0x29BF486A},
        {0x3BDD1190, 0x49372802},
        {0x26955F68, 0x35AF609A},
        {0x164D5E40, 0x4F275232},
        {0x6B056E18, 0x759F5CCA},
        {0x004BD6EF, 0x09176062},
        {0x480D3900, 0x6EE762F2},
        {0x437540C8, 0x698F3CFA},
        {0x072D43A0, 0x77075292},
        {0x02FE5577, 0x8117F12A},
        {0x1D9D5C50, 0x18F728C2},
        {0x30553228, 0x6D6F295A},
        {0x01234567, 0x89ABCDEF},
        {0x01234567, 0x89ABCDEF},
        {0x01234567, 0x89ABCDEF},
        {0xFFFFFFFF, 0xFFFFFFFF},
        {0x00000000, 0x00000000},
        {0x00000000, 0x00000000},
        {0xFFFFFFFF, 0xFFFFFFFF}
    };

    uint32_t ciphertexts[34][2] = {{0x4EF99745, 0x6198DD78},
        {0x51866FD5, 0xB85ECB8A},
        {0x7D856F9A, 0x613063F2},
        {0x2466DD87, 0x8B963C9D},
        {0x61F9C380, 0x2281B096},
        {0x7D0CC630, 0xAFDA1EC7},
        {0x4EF99745, 0x6198DD78},
        {0x0ACEAB0F, 0xC6A0A28D},
        {0x59C68245, 0xEB05282B},
        {0xB1B8CC0B, 0x250F09A0},
        {0x1730E577, 0x8BEA1DA4},
        {0xA25E7856, 0xCF2651EB},
        {0x353882B1, 0x09CE8F1A},
        {0x48F4D088, 0x4C379918},
        {0x432193B7, 0x8951FC98},
        {0x13F04154, 0xD69D1AE5},
        {0x2EEDDA93, 0xFFD39C79},
        {0xD887E039, 0x3C2DA6E3},
        {0x5F99D04F, 0x5B163969},
        {0x4A057A3B, 0x24D3977B},
        {0x452031C1, 0xE4FADA8E},
        {0x7555AE39, 0xF59B87BD},
        {0x53C55F9C, 0xB49FC019},
        {0x7A8E7BFA, 0x937E89A3},
        {0xCF9C5D7A, 0x4986ADB5},
        {0xD1ABB290, 0x658BC778},
        {0x55CB3774, 0xD13EF201},
        {0xFA34EC48, 0x47B268B2},
        {0xA7907951, 0x08EA3CAE},
        {0xC39E072D, 0x9FAC631D},
        {0x014933E0, 0xCDAFF6E4},
        {0xF21E9A77, 0xB71C49BC},
        {0x24594688, 0x5754369A},
        {0x6B5C5A9C, 0x5D9E0A5A}
    };

    int i = 0;

    std::cout << "Starting test case: ecb @ schneier.cc..." << std::endl;

    for (i = 0; i < 34; i++) {
        blowfish* ciph = new blowfish(keys[i], 8);
        uint32_t l = plaintexts[i][0];
        uint32_t r = plaintexts[i][1];
        ciph->encrypt(&l, &r);
        if (l != ciphertexts[i][0]) {
            std::cerr << "Error with test case: " << i << "::" << l << "::" <<
                      ciphertexts[i][0] << std::endl;
            return false;
        } else if (r != ciphertexts[i][1]) {
            std::cerr << "Error with test case: " << i << "::" << r << "::" <<
                      ciphertexts[i][1] << std::endl;
            return false;
        }

        ciph->decrypt(&l, &r);
        if (l != plaintexts[i][0]) {
            std::cerr << "Error with test case: " << i << "::" << l << "::" <<
                      plaintexts[i][0] << std::endl;
            return false;
        } else if (r != plaintexts[i][1]) {
            std::cerr << "Error with test case: " << i << "::" << r << "::" <<
                      plaintexts[i][1] << std::endl;
            return false;
        }

        delete ciph;
    }

    std::cout << "Passed test case: ecb @ schneier.cc" << std::endl;

    return true;
}

bool schneier_set_key_test()
{
    uint32_t plaintext_l = 0xFEDCBA98;
    uint32_t plaintext_r = 0x76543210;
    char key[24] = {(char) 0xF0, (char) 0xE1, (char) 0xD2, (char) 0xC3, (char) 0xB4, (char) 0xA5, (char) 0x96, (char) 0x87, (char) 0x78, (char) 0x69, (char) 0x5A, (char) 0x4B, (char) 0x3C, (char) 0x2D, (char) 0x1E, (char) 0x0F, (char) 0x00, (char) 0x11, (char) 0x22, (char) 0x33, (char) 0x44, (char) 0x55, (char) 0x66, (char) 0x77};

    uint32_t ciphertexts[24][2] = {{0xF9AD597C, 0x49DB005E},
        {0xE91D21C1, 0xD961A6D6},
        {0xE9C2B70A, 0x1BC65CF3},
        {0xBE1E6394, 0x08640F05},
        {0xB39E4448, 0x1BDB1E6E},
        {0x9457AA83, 0xB1928C0D},
        {0x8BB77032, 0xF960629D},
        {0xE87A244E, 0x2CC85E82},
        {0x15750E7A, 0x4F4EC577},
        {0x122BA70B, 0x3AB64AE0},
        {0x3A833C9A, 0xFFC537F6},
        {0x9409DA87, 0xA90F6BF2},
        {0x884F8062, 0x5060B8B4},
        {0x1F85031C, 0x19E11968},
        {0x79D9373A, 0x714CA34F},
        {0x93142887, 0xEE3BE15C},
        {0x03429E83, 0x8CE2D14B},
        {0xA4299E27, 0x469FF67B},
        {0xAFD5AED1, 0xC1BC96A8},
        {0x10851C0E, 0x3858DA9F},
        {0xE6F51ED7, 0x9B9DB21F},
        {0x64A6E14A, 0xFD36B46F},
        {0x80C7D7D4, 0x5A5479AD},
        {0x05044B62, 0xFA52D080}
    };

    int i = 1;

    std::cout << "Starting test case: set_key @ schneier.cc..." << std::endl;

    for (i = 0; i < 24; i++) {
        blowfish* ciph = new blowfish(key, i+1);
        uint32_t l = plaintext_l;
        uint32_t r = plaintext_r;

        ciph->encrypt(&l, &r);
        if (l != ciphertexts[i][0]) {
            std::cerr << "Error with test case: " << i << "::" << l << "::" <<
                      ciphertexts[i][0] << std::endl;
            return false;
        } else if (r != ciphertexts[i][1]) {
            std::cerr << "Error with test case: " << i << "::" << r << "::" <<
                      ciphertexts[i][1] << std::endl;
            return false;
        }

        ciph->decrypt(&l, &r);
        if (l != plaintext_l) {
            std::cerr << "Error with test case: " << i << "::" << l << "::" << plaintext_l
                      << std::endl;
            return false;
        } else if (r != plaintext_r) {
            std::cerr << "Error with test case: " << i << "::" << r << "::" << plaintext_r
                      << std::endl;
            return false;
        }

        delete ciph;
    }

    std::cout << "Passed test case: set_key @ schneier.cc" << std::endl;

    return true;
}


src/blowfish.cc
/*
 * Copyright (c) 2016 Alexander Scheel
 *
 * Implementation for the blowfish cryptographic primitive
*/

#include "blowfish.hh"
#include "blowfish_constants.hh"

#include <cstdlib>
#include <string>
#include <iostream>


/**
 * Initialize the blowfish sub-keys and s-boxes based off the given key.
 *
 * See /docs/src/blowfish.txt for more details.
**/
blowfish::blowfish(std::string key) : blowfish(key.c_str(), key.length())
{
}


/**
 * Initialize the blowfish sub-keys and s-boxes based off the given key.
 *
 * See /docs/src/blowfish.txt for more details.
**/
blowfish::blowfish(const char* key_data, int key_length)
{
    int i = 0;
    int key_loc = 0;

    uint32_t zero_left = 0;
    uint32_t zero_right = 0;

    uint8_t a;
    uint8_t b;
    uint8_t c;
    uint8_t d;

    uint32_t key_segment;

    // Initialize with digits of Pi
    for (i = 0; i < 18; i++) {
        this->subkeys[i] = blowfish_init_subkeys[i];
    }
    for (i = 0; i < 256; i++) {
        this->sboxes[0][i] = blowfish_init_sbox_0[i];
    }
    for (i = 0; i < 256; i++) {
        this->sboxes[1][i] = blowfish_init_sbox_1[i];
    }
    for (i = 0; i < 256; i++) {
        this->sboxes[2][i] = blowfish_init_sbox_2[i];
    }
    for (i = 0; i < 256; i++) {
        this->sboxes[3][i] = blowfish_init_sbox_3[i];
    }

    // xor all sub-keys with key segments.
    for (i = 0; i < 18; i++) {
        a = key_data[(key_loc+0) % key_length];
        b = key_data[(key_loc+1) % key_length];
        c = key_data[(key_loc+2) % key_length];
        d = key_data[(key_loc+3) % key_length];
        key_segment = (a << 24) + (b << 16) + (c << 8) + d;

        this->subkeys[i] = this->subkeys[i] ^ key_segment;

        key_loc = (key_loc+4) % key_length;
    }

    // Beginning with zero vectors, update sub-keys and s-boxes with encrypted
    // forms; ensures key gets propegated to all parts of the cipher
    for (i = 0; i < 18; i+=2) {
        this->encrypt(&zero_left, &zero_right);
        this->subkeys[i] = zero_left;
        this->subkeys[i+1] = zero_right;
    }

    for (i = 0; i < 256; i+=2) {
        this->encrypt(&zero_left, &zero_right);
        this->sboxes[0][i] = zero_left;
        this->sboxes[0][i+1] = zero_right;
    }

    for (i = 0; i < 256; i+=2) {
        this->encrypt(&zero_left, &zero_right);
        this->sboxes[1][i] = zero_left;
        this->sboxes[1][i+1] = zero_right;
    }

    for (i = 0; i < 256; i+=2) {
        this->encrypt(&zero_left, &zero_right);
        this->sboxes[2][i] = zero_left;
        this->sboxes[2][i+1] = zero_right;
    }

    for (i = 0; i < 256; i+=2) {
        this->encrypt(&zero_left, &zero_right);
        this->sboxes[3][i] = zero_left;
        this->sboxes[3][i+1] = zero_right;
    }
}


/**
 * Private function. Implements the Feistel function from the blowfish
 * specification; this is where the s-boxes are used.
 *
 * See /docs/src/blowfish.txt for more details.
**/
uint32_t blowfish::function_f(uint32_t data)
{
    // Set up each part
    uint8_t a = data >> 24;
    uint8_t b = data >> 16;
    uint8_t c = data >> 8;
    uint8_t d = data;

    // Feistel function; ensure 32-bit unsigned integer
    uint32_t result = ((this->sboxes[0][a] + this->sboxes[1][b]) ^
                       (this->sboxes[2][c])) + this->sboxes[3][d];

    return result;
}


/**
 * Encrypt either 8 bytes of data or two 4-byte integer halves.
 *
 * See /docs/src/blowfish.txt for more details.
**/
void blowfish::encrypt(uint32_t* left, uint32_t* right)
{
    uint32_t l = *left;
    uint32_t r = *right;

    /*
        With sub-keys i=0...15:
            left = left xor subkey[i]
            right = right xor f(left)
            swap left, right
        ~~
        Fully unrolled below.
    */
    l = l ^ this->subkeys[0];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[1];
    l = l ^ this->function_f(r);
    l = l ^ this->subkeys[2];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[3];
    l = l ^ this->function_f(r);
    l = l ^ this->subkeys[4];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[5];
    l = l ^ this->function_f(r);
    l = l ^ this->subkeys[6];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[7];
    l = l ^ this->function_f(r);
    l = l ^ this->subkeys[8];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[9];
    l = l ^ this->function_f(r);
    l = l ^ this->subkeys[10];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[11];
    l = l ^ this->function_f(r);
    l = l ^ this->subkeys[12];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[13];
    l = l ^ this->function_f(r);
    l = l ^ this->subkeys[14];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[15];
    l = l ^ this->function_f(r);

    /*
        swap left, right
        right = right xor subkey[16]
        left = left xor subkey[17]
        ~~
        Or don't bother swapping and swap on assignment.
    */
    l = l ^ this->subkeys[16];
    r = r ^ this->subkeys[17];

    *right = l;
    *left = r;
}


/**
 * Encrypt either 8 bytes of data or two 4-byte integer halves.
 * char* result must be freed after use, uint32_t* need not be.
 *
 * See /docs/src/blowfish.txt for more details.
**/
char* blowfish::encrypt(const char* data)
{
    uint32_t left;
    uint32_t right;
    char* result;

    result = (char *) malloc(sizeof(char) * 8);

    left = ((uint8_t) data[0] << 24) + ((uint8_t) data[1] << 16) + ((
                uint8_t) data[2] << 8) + (uint8_t) data[3];
    right = ((uint8_t) data[4] << 24) + ((uint8_t) data[5] << 16) + ((
                uint8_t) data[6] << 8) + (uint8_t) data[7];

    this->encrypt(&left, &right);

    result[0] = (left & 0xFF000000) >> 24;
    result[1] = (left & 0x00FF0000) >> 16;
    result[2] = (left & 0x0000FF00) >> 8;
    result[3] = left & 0x000000FF;
    result[4] = (right & 0xFF000000) >> 24;
    result[5] = (right & 0x00FF0000) >> 16;
    result[6] = (right & 0x0000FF00) >> 8;
    result[7] = right & 0x000000FF;

    return result;

}

std::string blowfish::encrypt_str(std::string data)
{
    if (data.length() != 8) {
        return "";
    }

    char* result = this->encrypt(data.c_str());

    // Construct a string with length equal to the original data's length
    std::string str_result(result, 8);
    free(result);

    return str_result;
}


/**
 * Decrypt either 8 bytes of data or two 4-byte integer halves.
 *
 * See /docs/src/blowfish.txt for more details.
**/
void blowfish::decrypt(uint32_t* left, uint32_t* right)
{
    uint32_t l = *left;
    uint32_t r = *right;

    /*
        With sub-keys i=17...2:
            left = left xor subkey[i]
            right = right xor f(left)
            swap left, right
        ~~
        Fully unrolled below.
    */
    l = l ^ this->subkeys[17];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[16];
    l = l ^ this->function_f(r);
    l = l ^ this->subkeys[15];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[14];
    l = l ^ this->function_f(r);
    l = l ^ this->subkeys[13];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[12];
    l = l ^ this->function_f(r);
    l = l ^ this->subkeys[11];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[10];
    l = l ^ this->function_f(r);
    l = l ^ this->subkeys[9];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[8];
    l = l ^ this->function_f(r);
    l = l ^ this->subkeys[7];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[6];
    l = l ^ this->function_f(r);
    l = l ^ this->subkeys[5];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[4];
    l = l ^ this->function_f(r);
    l = l ^ this->subkeys[3];
    r = r ^ this->function_f(l);
    r = r ^ this->subkeys[2];
    l = l ^ this->function_f(r);

    /*
        swap left, right
        right = right xor subkey[1]
        left = left xor subkey[0]
        ~~
        Or don't bother swapping and swap on assignment.
    */
    l = l ^ this->subkeys[1];
    r = r ^ this->subkeys[0];

    *right = l;
    *left = r;
}


/**
 * Decrypt either 8 bytes of data or two 4-byte integer halves.
 * char* pointer must be freed after use.
 *
 * See /docs/src/blowfish.txt for more details.
**/
char* blowfish::decrypt(const char* data)
{
    uint32_t left;
    uint32_t right;
    char* result;

    result = (char *) malloc(sizeof(char) * 8);

    left = ((uint8_t) data[0] << 24) + ((uint8_t) data[1] << 16) + ((
                uint8_t) data[2] << 8) + (uint8_t) data[3];
    right = ((uint8_t) data[4] << 24) + ((uint8_t) data[5] << 16) + ((
                uint8_t) data[6] << 8) + (uint8_t) data[7];

    this->decrypt(&left, &right);

    result[0] = (left & 0xFF000000) >> 24;
    result[1] = (left & 0x00FF0000) >> 16;
    result[2] = (left & 0x0000FF00) >> 8;
    result[3] = left & 0x000000FF;
    result[4] = (right & 0xFF000000) >> 24;
    result[5] = (right & 0x00FF0000) >> 16;
    result[6] = (right & 0x0000FF00) >> 8;
    result[7] = right & 0x000000FF;

    return result;
}

std::string blowfish::decrypt_str(std::string data)
{
    if (data.length() != 8) {
        return "";
    }

    char* result = this->decrypt(data.c_str());

    // Construct a string with length equal to the original data's length
    std::string str_result(result, 8);
    free(result);

    return str_result;
}
